In [2]:
import pandas as pd
pd.set_option('display.float_format', lambda x: '%.3f' % x)

# Python for Data Analysts
**Inclass for Week 1**
_______
## Objective Course
- Introduction to the `pandas` library. 
- Introduction to `DataFrame`  
- Data Types
- Exploratory Data Analysis I
- Indexing and Subsetting

Shortcut pada Jupyter Notebook:
- a : insert cell above
- b : insert cell below
- m : ubah cell ke markdown
- y : ubah cell ke code
- dd : delete cell
- ctrl+enter : run cell

## Keyword and Variable
Variable adalah nama yang dipakai untuk menunjukkan sebuah nilai. Tanda `=` dipakai untuk membuat variable baru.

In [2]:
# Code here
algoritma = "programming"
print(algoritma)


programming


Python adalah bahasa pemrograman yang *case-sensitive* sehingga penamaan antar variable menjadi hal yang perlu diperhatikan. Misalnya penulisan variable `activity` dengan awalan huruf `a` kecil berbeda dengan yang diawali dengan huruf `A` besar.

In [3]:
## Code here
algoritma == Algoritma


NameError: name 'Algoritma' is not defined

Berikut beberapa anjuran dalam memberikan nama variabel/identifier pada Python:
- Penamaan bisa menggunakan kombinasi dari huruf kapital (A-Z), huruf nomina (a-z), angka (0-9).
- Penamaan tidak boleh menggunakan angka di awal.
- Penamaan variabel tidak boleh menggunakan keyword pada Python
- Nama variabel bersifat case-sensitive sehingga penamaan variabel algoritma, ALGORITMA dan Algoritma adalah 3 nama variabel yang berbeda
- Karakter !, $ , &, dll tidak bisa digunakan dalam penamaan variabel.

Sedangkan keyword adalah kata kunci yang sudah di tetapkan oleh Python sebagai nama yang tidak bisa dipakai baik untuk penamaan fungsi, variabel, dan lainnya. Keyword ditulis dalam lower-case (huruf kecil semua) kecuali keyword True, False, dan None. Sejauh ini keyword yang ada pada Python adalah sebagai berikut:

`True`, `False`, `None`, `and`, `as`, `assert`, `break`, `class`, `continue`, `def`, `del`, `elif`, `else`
`except`, `finally`, `for`, `from`, `global`, `if`, `import`, `in`, `is`, `lambda`, `nonlocal`, `not`, `or`, `pass`
`raise`, `return`, `try`, `while`, `with`, `yield`

In [6]:
# Cek daftar keyword
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

Untuk membuktikan bahwa keyword tidak dapat digunakan untuk membuat nama variabel baru, cobalah untuk menimpan nilai 1 pada variabel `True`.

In [7]:
## Code here
True = 1

SyntaxError: can't assign to keyword (<ipython-input-7-b335ad4b97ec>, line 2)

Bagaimana? Apa yang Anda temukan? True tidak bisa di assign dengan sebuah nilai karena dia adalah sebuah keyword.

## Introduction to Pandas Library
Pandas adalah library yang cukup powerfull sebagai tools analisis data dan struktur pada Python. Dengan Pandas kita dapat mengolah data dengan mudah. Pandas menyediakan salah satu fitur bernama Dataframe. Dengan dataframe kita dapat membaca sebuah file, mengolah suatu data dengan menggunakan operasi seperti join, distinct, group by, agregasi, dan teknik lainnya. Untuk menggunakan Pandas, kita perlu import terlebih dahulu librarynya dengan cara berikut ini:

In [8]:
import pandas
print(pandas.__version__)

0.24.2


In [9]:
import pandas as pd
print(pd.__version__)

0.24.2


Perlu diperhatikan bahwa semua fungsi pada pandas dapat diakses dengan cara dipanggil seperti syntax : `pandas.function_name()`. Langkah pertama yang akan kita lakukan adalah membaca data. Anda dapat menggunakan `.read_csv()` untuk membaca sebuah file dengan format `.csv`.

In [4]:
rice1 = pd.read_csv("data_input/rice.csv")
rice1.head()

Unnamed: 0  receipt_id  receipts_item_id    purchase_time category  \
0           1     9622257          32369294  7/22/2018 21:19     Rice   
1           2     9446359          31885876  7/15/2018 16:17     Rice   
2           3     9470290          31930241  7/15/2018 12:12     Rice   
3           4     9643416          32418582   7/24/2018 8:27     Rice   
4           5     9692093          32561236  7/26/2018 11:28     Rice   

  sub_category       format  unit_price  discount  quantity yearmonth  
0         Rice  supermarket  128000.000         0         1   2018-07  
1         Rice   minimarket  102750.000         0         1   2018-07  
2         Rice  supermarket   64000.000         0         3   2018-07  
3         Rice   minimarket   65000.000         0         1   2018-07  
4         Rice  supermarket  124500.000         0         1   2018-07

In [5]:
# code here
rice = pd.read_csv("data_input/rice.csv", index_col=0)
rice.head()

receipt_id  receipts_item_id    purchase_time category sub_category  \
1     9622257          32369294  7/22/2018 21:19     Rice         Rice   
2     9446359          31885876  7/15/2018 16:17     Rice         Rice   
3     9470290          31930241  7/15/2018 12:12     Rice         Rice   
4     9643416          32418582   7/24/2018 8:27     Rice         Rice   
5     9692093          32561236  7/26/2018 11:28     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket  128000.000         0         1   2018-07  
2   minimarket  102750.000         0         1   2018-07  
3  supermarket   64000.000         0         3   2018-07  
4   minimarket   65000.000         0         1   2018-07  
5  supermarket  124500.000         0         1   2018-07

Dokumentasi `read_csv()` selengkapnya ada [disini](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)

### Konwledge Check
Referring to the previous **Python keywords** concept. Which of the following 4 lines of code will evaluate without raising an error?

- [ ] `pd.read_csv("data_input/rice.csv", index_col=false)`
- [ ] `Import pandas as pd`
- [x] `print(100-2)`
- [ ] `None = 2`

In [ ]:
## Code here
print(100-2)



### Dive Deeper
Let's dive deeper into understanding the `index_col` parameter. From the documentation:

> `index_col` : int or sequence or `False`  
Column to use as the row labels of the DataFrame.

1. How would you change the `read.csv()` code such that the DataFrame uses `receipt_id` as the row label (index)? 
2. `pandas.DataFrame.head()` accepts an additional parameter, `n`, and returns the first `n` rows of the DataFrame; Set `n=8` to see the first 8 rows of your `rice` DataFrame
3. The opposite of `.head()` is `.tail()`. It returns the last `n` row of your DataFrame. Create a new cell below and print the last 4 rows of our DataFrame 

In [6]:
## Code here
rice2 = pd.read_csv("data_input/rice.csv",index_col='receipt_id')
rice2.head(8)

Unnamed: 0  receipts_item_id    purchase_time category  \
receipt_id                                                           
9622257              1          32369294  7/22/2018 21:19     Rice   
9446359              2          31885876  7/15/2018 16:17     Rice   
9470290              3          31930241  7/15/2018 12:12     Rice   
9643416              4          32418582   7/24/2018 8:27     Rice   
9692093              5          32561236  7/26/2018 11:28     Rice   
9504155              6          32030785  7/17/2018 18:05     Rice   
9822589              7          32935097  7/29/2018 18:18     Rice   
9706953              8          32593606  7/25/2018 12:48     Rice   

           sub_category       format  unit_price  discount  quantity yearmonth  
receipt_id                                                                      
9622257            Rice  supermarket  128000.000         0         1   2018-07  
9446359            Rice   minimarket  102750.000         0         1   2018-07  
9470290            Rice  supermarket   64000.000         0         3   2018-07  
9643416            Rice   minimarket   65000.000         0         1   2018-07  
9692093            Rice  supermarket  124500.000         0         1   2018-07  
9504155            Rice   minimarket   63500.000         0         1   2018-07  
9822589            Rice  supermarket   66500.000         0         1   2018-07  
9706953            Rice   minimarket   62500.000         0         1   2018-07

In [15]:
rice2.tail(4)

Unnamed: 0  receipts_item_id     purchase_time category  \
receipt_id                                                            
5598782          11997          16999147   12/2/2017 14:12     Rice   
5735850          11998          17434503  12/13/2017 19:17     Rice   
5678748          11999          17317935   12/8/2017 22:04     Rice   
5702411          12000          17341559  12/10/2017 17:04     Rice   

           sub_category       format  unit_price  discount  quantity yearmonth  
receipt_id                                                                      
5598782            Rice   minimarket     64000.0         0         1   2017-12  
5735850            Rice  supermarket     59990.0      3000         1   2017-12  
5678748            Rice   minimarket     59500.0      3000         1   2017-12  
5702411            Rice   minimarket     59500.0      3000         1   2017-12

## Data Types


- Cek tipe data : `dtypes`

In [17]:
# method
rice.head()

receipt_id  receipts_item_id    purchase_time category sub_category  \
1     9622257          32369294  7/22/2018 21:19     Rice         Rice   
2     9446359          31885876  7/15/2018 16:17     Rice         Rice   
3     9470290          31930241  7/15/2018 12:12     Rice         Rice   
4     9643416          32418582   7/24/2018 8:27     Rice         Rice   
5     9692093          32561236  7/26/2018 11:28     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket    128000.0         0         1   2018-07  
2   minimarket    102750.0         0         1   2018-07  
3  supermarket     64000.0         0         3   2018-07  
4   minimarket     65000.0         0         1   2018-07  
5  supermarket    124500.0         0         1   2018-07

In [19]:
# cek tipe data rice
#atribut
rice.dtypes



receipt_id            int64
receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
yearmonth            object
dtype: object

In [34]:
rice.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12000 entries, 1 to 12000
Data columns (total 10 columns):
receipt_id          12000 non-null int64
receipts_item_id    12000 non-null int64
purchase_time       12000 non-null object
category            12000 non-null object
sub_category        12000 non-null object
format              12000 non-null object
unit_price          12000 non-null float64
discount            12000 non-null int64
quantity            12000 non-null int64
yearmonth           12000 non-null object
dtypes: float64(1), int64(4), object(5)
memory usage: 1.0+ MB


### Knowledge check: `.dtypes` and pandas attributes
Look at the following code - what is the expected output from the following code? Why?
```
x = [2019, 4, 'data science']
x.dtypes
```

Hint: Try `type(x)` and verify the type for object `x`.

In [20]:
#code here
x = [2019, 4, 'data science']

# x.dtypes
type(x)


list

In [22]:
type(rice)

pandas.core.frame.DataFrame

In [21]:
x

[2019, 4, 'data science']

Atribut `dtypes` hanya dimiliki oleh objek dataframe oleh karena itu `dtypes` tidak bisa diaplikasikan untuk list.

Perhatikan dataframe di bawah ini untuk contoh lain dari tipe data pada pandas!

In [23]:
employees = pd.DataFrame({
    'name': ['Anita', 'Brian'],
    'age': [34, 29],
    'joined': [pd.Timestamp('20190410'), pd.Timestamp('20171128')],
    'degree': [True, False],
    'hourlyrate': [35.5, 29],
    'division': ['HR', 'Product']
})
employees.dtypes

name                  object
age                    int64
joined        datetime64[ns]
degree                  bool
hourlyrate           float64
division              object
dtype: object

In [68]:
age = pd.Series([56,34,78,23])
type(age)

pandas.core.series.Series

In [69]:
age

0    56
1    34
2    78
3    23
dtype: int64

In [70]:
sales = [2000000, "Fafilia", "Algoritma"]
type(sales)

list

List : bisa terdiri dari berbagai macam tipe data
Series: hanya terdiri dari 1 tipe data 

- `name` [`object`]: store text values
- `age` [`int`]: integer values
- `joined` [`datetime`]: date and time values
- `degree` [`bool`]: True/False values
- `hourlyrate` [`float`]: floating point values
- `division` [`object`]: store text values

### Categorical and Numerical Variables
Tipe data categorical adalah tipe data yang memiliki karakteristik dimana nilainya selalu berulang pada sebuah kolom di dataframe. Ketika kita bekerja dengan variabel kategorik, ada 2 alasan mengapa kita harus menggunakan tipe data categorical pada pandas. Alasan pertama dengan mempertimbangkan "bussiness prespective". Hal ini dapat menginformasikan dan memandu seorang Analyst, pada pertanyaan seperti metode statistik atau tipe plot mana yang digunakan untuk mengolah data. 

Alasan kedua adalah dengan melihat dari sisi technicalnya. Ketika kita bekerja dengan tipe data categorical pada pandas, hal ini akan jauh menghemat memori dan tentunya akan menambah kecepan komputasional.

Berikut karakteristik Categorical variabel yang dikutip dari official page:
> Categoricals are a pandas data type corresponding to categorical variables in statistics. A categorical variable takes on a limited, and usually fixed, number of possible values (categories; levels in R). Examples are gender, social class, blood type, country affiliation or rating via Likert scales.

Untuk mengubah tipe data ke categorical pada pandas, Anda dapat melakukannya dengan method `astype()` berikut:

**Formula** 
```
df['column_name'] = df['column_name'].astype('new_data_types')
```
**Contoh**
```
employees['marital_status'] = employees['marital_status'].astype('category')
```

Perhatikan kembali dataframe `employees`. Adakah kolom yang tidak tersimpan dengan tipe data yang benar? Ubahlah ke tipe data yang tepat dengan menggunakan method `astype()`!

In [76]:
# code here

employees['division'] = employees['division'].astype('category')
employees.dtypes

name                  object
age                    int64
joined        datetime64[ns]
degree                  bool
hourlyrate           float64
division            category
dtype: object

- Categorical -> category
- Numerical -> integer (int64), float(float64)

In [80]:
employees[['division','name']] = employees[['division','name']].astype('category')

In [81]:
employees.dtypes

name                category
age                    int64
joined        datetime64[ns]
degree                  bool
hourlyrate           float64
division            category
dtype: object

## Exploratory Data Analysis I
Exploratory Data Analysis (EDA) mengacu pada proses melakukan investigasi awal pada data, seringkali dengan tujuan untuk lebih mengenal dengan karakteristik data tertentu. EDA biasanya dilakukan dengan bantuan ringkasan statistik dan teknik grafis sederhana untuk melihat struktur data yang kita miliki.

Beberapa tools sederhana pada Pandas yang dapat digunakan untuk melakukan EDA adalah sebagai berikut:
- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`

- head() dan tail()

Method `head()` digunakan untuk inspeksi 5 data teratas dan `tail()` digunakan untuk inspeksi 5 data terbawah dari sebuah dataframe

In [82]:
# code here
rice.head(3)


receipt_id  receipts_item_id    purchase_time category sub_category  \
1     9622257          32369294  7/22/2018 21:19     Rice         Rice   
2     9446359          31885876  7/15/2018 16:17     Rice         Rice   
3     9470290          31930241  7/15/2018 12:12     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket    128000.0         0         1   2018-07  
2   minimarket    102750.0         0         1   2018-07  
3  supermarket     64000.0         0         3   2018-07

In [83]:
rice.tail(3)

receipt_id  receipts_item_id     purchase_time category sub_category  \
11998     5735850          17434503  12/13/2017 19:17     Rice         Rice   
11999     5678748          17317935   12/8/2017 22:04     Rice         Rice   
12000     5702411          17341559  12/10/2017 17:04     Rice         Rice   

            format  unit_price  discount  quantity yearmonth  
11998  supermarket     59990.0      3000         1   2017-12  
11999   minimarket     59500.0      3000         1   2017-12  
12000   minimarket     59500.0      3000         1   2017-12

- describe()

Method `describe()` menampilkan 8 ringkasan statistika yang dapat membantu kita untuk melihat descriptive statistic dari seluruh kolom numerik pada dataframe(default). Ringkasan statistika yang dimaksud adalah sebagai berikut:
- Count : banyaknya data pada dataframe
- Mean : rata-rata nilai dari sebuah kolom di dataframe
- Standard Deviation : jarak anatara data ke titik pemusatan
- Minimum Value : nilai terkecil dari keseluruhan data
- 25th Percentile (Q1)
- 50th Percentile (Q2/Median)
- 75th Percentile (Q3)
- Maximum Value : nilai terbesar dari keseluruhan data

In [33]:
# code here
rice.describe()

receipt_id  receipts_item_id  unit_price   discount  quantity
count    12000.000         12000.000   12000.000  12000.000 12000.000
mean   7650135.157      24579495.813   70013.146    835.306     1.333
std    1873838.101       7171105.455   29905.391   6207.476     0.980
min    3173994.000       9282023.000    9395.000  -4100.000     1.000
25%    5983209.250      18206938.000   61900.000      0.000     1.000
50%    7443618.000      22343373.500   63500.000      0.000     1.000
75%    9149786.000      31083788.000   66000.000      0.000     1.000
max   11463208.000      38429394.000  219400.000 320000.000    19.000

In [85]:
rice['receipt_id'].count()

12000

In [88]:
rice['unit_price'].mean()

70013.14631275

Misalkan di unit price mean = 70000, std = 30000. Ketika saya punya data baru 120000

In [89]:
rice['unit_price'].std()

29905.39143678175

In [90]:
rice['unit_price'].min()

9395.0

In [93]:
rice['unit_price'].quantile(0.75)

66000.0

Secara default yang ditampilkan oleh method `quantile()` adalah nilai Q2/Median.

In [94]:
rice['unit_price'].max()

219400.0

In [ ]:
Untuk melihat statistika deskripsi dari variabel non-numeric bisa dengancara di bawah ini:

In [117]:
rice2= pd.read_csv('data_input/rice.csv', index_col=0)
rice2.describe(exclude='number')

purchase_time category sub_category      format yearmonth
count            12000    12000        12000       12000     12000
unique           11609        1            1           3        12
top     6/14/2018 0:00     Rice         Rice  minimarket   2018-05
freq                 4    12000        12000        7088      1000

In [116]:
rice2.describe(include='object')

purchase_time category sub_category      format yearmonth
count            12000    12000        12000       12000     12000
unique           11609        1            1           3        12
top     6/14/2018 0:00     Rice         Rice  minimarket   2018-05
freq                 4    12000        12000        7088      1000

In [7]:
rice.describe(exclude='object')

receipt_id  receipts_item_id  unit_price   discount  quantity
count    12000.000         12000.000   12000.000  12000.000 12000.000
mean   7650135.157      24579495.813   70013.146    835.306     1.333
std    1873838.101       7171105.455   29905.391   6207.476     0.980
min    3173994.000       9282023.000    9395.000  -4100.000     1.000
25%    5983209.250      18206938.000   61900.000      0.000     1.000
50%    7443618.000      22343373.500   63500.000      0.000     1.000
75%    9149786.000      31083788.000   66000.000      0.000     1.000
max   11463208.000      38429394.000  219400.000 320000.000    19.000

In [97]:
rice.describe(include='object')

purchase_time category sub_category      format yearmonth
count            12000    12000        12000       12000     12000
unique           11609        1            1           3        12
top     6/14/2018 0:00     Rice         Rice  minimarket   2018-05
freq                 4    12000        12000        7088      1000

- unique : nilai unik di setiap kolom atau variabel
- top : nilai terbanyak (modus)
- freq : banyaknya kemunculan modus

In [35]:
rice['format'].unique()

array(['supermarket', 'minimarket', 'hypermarket'], dtype=object)

In [36]:
rice['format'].nunique()

3

- shape dan size

`shape` dan `size` adalah atribut dari sebuah dataframe yang memberikan informasi terkait dimensi data dan ukuran data

In [98]:
# code here
rice.shape

(12000, 10)

In [ ]:
(12000, 10) : angka pertama menunjukkan row/observasi, angka kedua menunjukkan banyaknya kolom)

In [99]:
rice.size

120000

In [ ]:
size = row x kolom

- axes

`axes` adalah atribut dari sebuah dataframe yang memberikan informasi terkait index dataframe (baik index column maupun index baris)

In [100]:
# code here

rice.axes

[Int64Index([    1,     2,     3,     4,     5,     6,     7,     8,     9,
                10,
             ...
             11991, 11992, 11993, 11994, 11995, 11996, 11997, 11998, 11999,
             12000],
            dtype='int64', length=12000),
 Index(['receipt_id', 'receipts_item_id', 'purchase_time', 'category',
        'sub_category', 'format', 'unit_price', 'discount', 'quantity',
        'yearmonth'],
       dtype='object')]

In [101]:
rice.axes[0]

Int64Index([    1,     2,     3,     4,     5,     6,     7,     8,     9,
               10,
            ...
            11991, 11992, 11993, 11994, 11995, 11996, 11997, 11998, 11999,
            12000],
           dtype='int64', length=12000)

In [102]:
rice.axes[1]

Index(['receipt_id', 'receipts_item_id', 'purchase_time', 'category',
       'sub_category', 'format', 'unit_price', 'discount', 'quantity',
       'yearmonth'],
      dtype='object')

- axes[0] : melihat index baris
- axes[1] : melihat index kolom
    

- dtypes

Atribut `dytpes` digunakan untuk melakukan inspeksi tipe data

In [103]:
# code here
rice.dtypes


receipt_id            int64
receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
yearmonth            object
dtype: object

In [37]:
# rice data types
rice[['category','sub_category','format']] = rice[['category','sub_category','format']].astype('category')


In [38]:
rice['purchase_time'] = rice['purchase_time'].astype('datetime64')

In [39]:
rice['unit_price'] = rice['unit_price'].astype('int64')

In [112]:
rice.dtypes

receipt_id                   int64
receipts_item_id             int64
purchase_time       datetime64[ns]
category                  category
sub_category              category
format                    category
unit_price                   int64
discount                     int64
quantity                     int64
yearmonth           datetime64[ns]
year_month          datetime64[ns]
dtype: object

### Knowledge Check: Data types

Supposed we have a pandas DataFrame named `inventory`. 

1. We called `inventory.dtypes` and got the following output. Which of the column likely require type conversion because it seems to have the wrong data type? Choose all that apply.

    - [ ] `units_instock`: int64
    - [ ] `discount_price`: float64
    - [ ] `item_name`: object
    - [x] `units_sold`: object
    

2. We would like to know the number of columns in `inventory`. Which of the following code would print the number of columns in `inventory`? Choose all that apply.

    - [x] `print(len(inventory.columns))`
    - [x] `print(inventory.shape[1])`
    - [ ] `print(len(rice.axes[1]))`


In [120]:
# code here
inventory = pd.DataFrame({
    'units_instock' : [20,35,67],
    'discount_price' : [2.4,5.6,7.8],
    'item_name' : ['bawang','garam','gula'],
    'unit_sold' : ['289','675','980']
})
inventory.dtypes


units_instock       int64
discount_price    float64
item_name          object
unit_sold          object
dtype: object

In [121]:
inventory['unit_sold'] = inventory['unit_sold'].astype('int64')

In [124]:
print(len(inventory.columns))
print(inventory.shape[1])
print(len(rice.axes[1]))

4
4
11


## Indexing and Subsetting with Pandas

Indexing digunakan untuk memilih(filter) dan meringkas beberapa set data (sebagian data) yang merupakan bagian critical dari workflow data analysis yang sedang dikerjakan. Contohnya:
- Compare sales pada tahun 2018 vs 2019
- Identifikasi peluang penjualan pada segment pasar (ex : Wholesale vs Retail)
- Melihat quarter terbaik untuk setiap tahun yang dapat digunakan untuk tujuan promosi,dll
- dsb

Untuk melakukan indexing dan subseting (slicing data) pada Pandas, ada beberapa cara yang bisa digunakan. Method yang mendukung hal tersebut adalah sebagai berikut:
- `head()` and `tail()`  
- `select_dtypes()`  
- Using `.drop()` 
- The `[]` operator
- `.loc`  
- `.iloc`
- Conditional subsetting

- **`select_dtypes()`**

Method `select_dtypes()` digunakan untuk memilih/memfilter kolom-kolom tertentu sesuai dengan nilai parameter yang ada di dalam method `select_dtypes()`. Ada 2 parameter yang dapat digunakan di dalam method `select_dtypes()` yaitu parameter `include` dan `exclude`. Misalnya jika Anda menempatkan parameter `include = 'category'` ini artinya Anda memilih semua kolom dengan tipe data 'category' yang ada di dataframe untuk ditampilkan. Sebaliknya jika Anda menempatkan parameter `exclude = 'category'` maka kolom-kolom yang akan ditampilkan adalah semua kolom dengan tipe data selain 'category'.

In [42]:
# Code here
rice.select_dtypes(include='category').describe()


category sub_category      format
count     12000        12000       12000
unique        1            1           3
top        Rice         Rice  minimarket
freq      12000        12000        7088

In [43]:
rice.select_dtypes(exclude='category').describe()

receipt_id  receipts_item_id  unit_price   discount  quantity
count    12000.000         12000.000   12000.000  12000.000 12000.000
mean   7650135.157      24579495.813   70013.146    835.306     1.333
std    1873838.101       7171105.455   29905.392   6207.476     0.980
min    3173994.000       9282023.000    9395.000  -4100.000     1.000
25%    5983209.250      18206938.000   61900.000      0.000     1.000
50%    7443618.000      22343373.500   63500.000      0.000     1.000
75%    9149786.000      31083788.000   66000.000      0.000     1.000
max   11463208.000      38429394.000  219400.000 320000.000    19.000

In [46]:
rice.describe(include='object')

yearmonth
count      12000
unique        12
top      2018-04
freq        1000

In [48]:
rice.select_dtypes(include='object').head()

yearmonth
1   2018-07
2   2018-07
3   2018-07
4   2018-07
5   2018-07

- describe() : untuk menampilkan ringkasan statistika dari sebuah dataframe
- select_dtypes() : untukuk memfilter/menyeleksi kolom-kolom dengan tipe data tertentu yang dideskripsikan sebelumnya

**Extra Challenge**

Bagaimana cara Anda untuk menyeleksi kolom dengan tipe data `integer` dan `float` untuk ditampilkan?

In [45]:
# code here
rice.select_dtypes(include=['int64','float64']).head()


receipt_id  receipts_item_id  unit_price  discount  quantity
1     9622257          32369294      128000         0         1
2     9446359          31885876      102750         0         1
3     9470290          31930241       64000         0         3
4     9643416          32418582       65000         0         1
5     9692093          32561236      124500         0         1

In [49]:
rice.select_dtypes(include='number').head()

receipt_id  receipts_item_id  unit_price  discount  quantity
1     9622257          32369294      128000         0         1
2     9446359          31885876      102750         0         1
3     9470290          31930241       64000         0         3
4     9643416          32418582       65000         0         1
5     9692093          32561236      124500         0         1

- **`drop()`**

Method `drop()` digunakan untuk menyeleksi baris atau kolom yang tidak ingin digunakan untuk tujuan analisis. Secara default method ini akan menghapus baris.

In [51]:
rice1.head()

Unnamed: 0  receipt_id  receipts_item_id    purchase_time category  \
0           1     9622257          32369294  7/22/2018 21:19     Rice   
1           2     9446359          31885876  7/15/2018 16:17     Rice   
2           3     9470290          31930241  7/15/2018 12:12     Rice   
3           4     9643416          32418582   7/24/2018 8:27     Rice   
4           5     9692093          32561236  7/26/2018 11:28     Rice   

  sub_category       format  unit_price  discount  quantity yearmonth  
0         Rice  supermarket  128000.000         0         1   2018-07  
1         Rice   minimarket  102750.000         0         1   2018-07  
2         Rice  supermarket   64000.000         0         3   2018-07  
3         Rice   minimarket   65000.000         0         1   2018-07  
4         Rice  supermarket  124500.000         0         1   2018-07

In [50]:
# code here
rice.head()

receipt_id  receipts_item_id       purchase_time category sub_category  \
1     9622257          32369294 2018-07-22 21:19:00     Rice         Rice   
2     9446359          31885876 2018-07-15 16:17:00     Rice         Rice   
3     9470290          31930241 2018-07-15 12:12:00     Rice         Rice   
4     9643416          32418582 2018-07-24 08:27:00     Rice         Rice   
5     9692093          32561236 2018-07-26 11:28:00     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket      128000         0         1   2018-07  
2   minimarket      102750         0         1   2018-07  
3  supermarket       64000         0         3   2018-07  
4   minimarket       65000         0         1   2018-07  
5  supermarket      124500         0         1   2018-07

In [8]:
rice.drop(2).head()

receipt_id  receipts_item_id    purchase_time category sub_category  \
1     9622257          32369294  7/22/2018 21:19     Rice         Rice   
3     9470290          31930241  7/15/2018 12:12     Rice         Rice   
4     9643416          32418582   7/24/2018 8:27     Rice         Rice   
5     9692093          32561236  7/26/2018 11:28     Rice         Rice   
6     9504155          32030785  7/17/2018 18:05     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket  128000.000         0         1   2018-07  
3  supermarket   64000.000         0         3   2018-07  
4   minimarket   65000.000         0         1   2018-07  
5  supermarket  124500.000         0         1   2018-07  
6   minimarket   63500.000         0         1   2018-07

In [69]:
rice.drop(columns=['receipt_id','receipts_item_id'],axis=1).head()

purchase_time category sub_category       format  unit_price  \
1 2018-07-22 21:19:00     Rice         Rice  supermarket      128000   
2 2018-07-15 16:17:00     Rice         Rice   minimarket      102750   
3 2018-07-15 12:12:00     Rice         Rice  supermarket       64000   
4 2018-07-24 08:27:00     Rice         Rice   minimarket       65000   
5 2018-07-26 11:28:00     Rice         Rice  supermarket      124500   

   discount  quantity yearmonth  
1         0         1   2018-07  
2         0         1   2018-07  
3         0         3   2018-07  
4         0         1   2018-07  
5         0         1   2018-07

In [56]:
rice.drop(columns='receipt_id').head()

receipts_item_id       purchase_time category sub_category       format  \
1          32369294 2018-07-22 21:19:00     Rice         Rice  supermarket   
2          31885876 2018-07-15 16:17:00     Rice         Rice   minimarket   
3          31930241 2018-07-15 12:12:00     Rice         Rice  supermarket   
4          32418582 2018-07-24 08:27:00     Rice         Rice   minimarket   
5          32561236 2018-07-26 11:28:00     Rice         Rice  supermarket   

   unit_price  discount  quantity yearmonth  
1      128000         0         1   2018-07  
2      102750         0         1   2018-07  
3       64000         0         3   2018-07  
4       65000         0         1   2018-07  
5      124500         0         1   2018-07

Syntax drop secara default : df.drop(index=,axis=0)

In [71]:
rice.index[7:12]

Int64Index([8, 9, 10, 11, 12], dtype='int64')

In [10]:
rice1.drop(index=rice1.index[7:12]).head(10)

Unnamed: 0  receipt_id  receipts_item_id    purchase_time category  \
0            1     9622257          32369294  7/22/2018 21:19     Rice   
1            2     9446359          31885876  7/15/2018 16:17     Rice   
2            3     9470290          31930241  7/15/2018 12:12     Rice   
3            4     9643416          32418582   7/24/2018 8:27     Rice   
4            5     9692093          32561236  7/26/2018 11:28     Rice   
5            6     9504155          32030785  7/17/2018 18:05     Rice   
6            7     9822589          32935097  7/29/2018 18:18     Rice   
12          13     9691090          32552145  7/26/2018 11:43     Rice   
13          14     9624752          32369065  7/23/2018 14:22     Rice   
14          15     9543709          32142234  7/19/2018 18:05     Rice   

   sub_category       format  unit_price  discount  quantity yearmonth  
0          Rice  supermarket  128000.000         0         1   2018-07  
1          Rice   minimarket  102750.000         0         1   2018-07  
2          Rice  supermarket   64000.000         0         3   2018-07  
3          Rice   minimarket   65000.000         0         1   2018-07  
4          Rice  supermarket  124500.000         0         1   2018-07  
5          Rice   minimarket   63500.000         0         1   2018-07  
6          Rice  supermarket   66500.000         0         1   2018-07  
12         Rice   minimarket   66500.000         0         1   2018-07  
13         Rice   minimarket   62900.000         0         1   2018-07  
14         Rice   minimarket   64000.000         0         1   2018-07

- **`[]` operator**

Digunakan untuk melakukan subseting dengan cara mengiris(slicing) baris pada dataframe. Formula penulisannya adalah `[start:end]` dengan mengikuti aturan indexing pada python (dimulai dari 0).

Dimana:
end = n-1

In [73]:
rice.axes[1]

Index(['receipt_id', 'receipts_item_id', 'purchase_time', 'category',
       'sub_category', 'format', 'unit_price', 'discount', 'quantity',
       'yearmonth'],
      dtype='object')

In [61]:
# code here

rice[5:10]

receipt_id  receipts_item_id       purchase_time category sub_category  \
6      9504155          32030785 2018-07-17 18:05:00     Rice         Rice   
7      9822589          32935097 2018-07-29 18:18:00     Rice         Rice   
8      9706953          32593606 2018-07-25 12:48:00     Rice         Rice   
9      9699516          32573843 2018-07-26 16:41:00     Rice         Rice   
10     9444092          31913062 2018-07-14 21:17:00     Rice         Rice   

         format  unit_price  discount  quantity yearmonth  
6    minimarket       63500         0         1   2018-07  
7   supermarket       66500         0         1   2018-07  
8    minimarket       62500         0         1   2018-07  
9    minimarket       62500         0         1   2018-07  
10  supermarket       64000         0         3   2018-07

In [ ]:
rice.drop(columns=['receipt_id','receipts_item_id'])

Catatan : Perbedaaan List dengan `[] operator`.
- List : biasanya tidak didahului oleh object ataupun atribut
Contoh, x = ['algoritma',23000,'A']
- `[] operator` untuk subsetting
Contoh, rice.axes[1] atau rice[7:12]

### Knowledge Check: Slicing

Recalling that the `end` is not inclusive and Python's 0-based indexing behavior, if we have wanted to subset the **8th to 12th** row of our data, how would we have done it instead? Pick the right answer and try it in a new code cell below.

- [ ] `rice[7:12]`
- [ ] `rice[8:12]`
- [ ] `rice[7:13]`
- [ ] `rice[8:13]`

In [62]:
#code here
rice[7:12]



receipt_id  receipts_item_id       purchase_time category sub_category  \
8      9706953          32593606 2018-07-25 12:48:00     Rice         Rice   
9      9699516          32573843 2018-07-26 16:41:00     Rice         Rice   
10     9444092          31913062 2018-07-14 21:17:00     Rice         Rice   
11     9198777          31125365 2018-07-02 15:39:00     Rice         Rice   
12     9763483          32856560 2018-07-31 05:51:00     Rice         Rice   

         format  unit_price  discount  quantity yearmonth  
8    minimarket       62500         0         1   2018-07  
9    minimarket       62500         0         1   2018-07  
10  supermarket       64000         0         3   2018-07  
11   minimarket      112500         0         1   2018-07  
12  hypermarket       77750         0         1   2018-07

- **`.loc` dan `iloc`**

Dengan menggunakan `.loc` dan `iloc` kita dapat melakukan pengirisan pada index baris dan kolom. PErbedaan yang mendasar dari kedua operator ini adalah:
- Ketika Anda menggunakan operasi `.loc` maka ini merujuk pada lokasi **nama** baris atau kolomnya
- Ketika Anda menggunakan operasi `.iloc` maka ini merujuk pada lokasi **index** baris atau kolomnya

In [ ]:
loc[0:kolom]

In [74]:
# code here
rice.head()


receipt_id  receipts_item_id       purchase_time category sub_category  \
1     9622257          32369294 2018-07-22 21:19:00     Rice         Rice   
2     9446359          31885876 2018-07-15 16:17:00     Rice         Rice   
3     9470290          31930241 2018-07-15 12:12:00     Rice         Rice   
4     9643416          32418582 2018-07-24 08:27:00     Rice         Rice   
5     9692093          32561236 2018-07-26 11:28:00     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket      128000         0         1   2018-07  
2   minimarket      102750         0         1   2018-07  
3  supermarket       64000         0         3   2018-07  
4   minimarket       65000         0         1   2018-07  
5  supermarket      124500         0         1   2018-07

In [ ]:
Kita akan menyeleksi/memilih baris 1-5 pada dataframe rice

Syntax : df.loc[baris,kolom] -> baris sifatnya inclusive (harus ada)

In [76]:
rice.loc[1:5,:]

receipt_id  receipts_item_id       purchase_time category sub_category  \
1     9622257          32369294 2018-07-22 21:19:00     Rice         Rice   
2     9446359          31885876 2018-07-15 16:17:00     Rice         Rice   
3     9470290          31930241 2018-07-15 12:12:00     Rice         Rice   
4     9643416          32418582 2018-07-24 08:27:00     Rice         Rice   
5     9692093          32561236 2018-07-26 11:28:00     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket      128000         0         1   2018-07  
2   minimarket      102750         0         1   2018-07  
3  supermarket       64000         0         3   2018-07  
4   minimarket       65000         0         1   2018-07  
5  supermarket      124500         0         1   2018-07

In [12]:
rice.loc[:,:].head()

receipt_id  receipts_item_id    purchase_time category sub_category  \
1     9622257          32369294  7/22/2018 21:19     Rice         Rice   
2     9446359          31885876  7/15/2018 16:17     Rice         Rice   
3     9470290          31930241  7/15/2018 12:12     Rice         Rice   
4     9643416          32418582   7/24/2018 8:27     Rice         Rice   
5     9692093          32561236  7/26/2018 11:28     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket  128000.000         0         1   2018-07  
2   minimarket  102750.000         0         1   2018-07  
3  supermarket   64000.000         0         3   2018-07  
4   minimarket   65000.000         0         1   2018-07  
5  supermarket  124500.000         0         1   2018-07

In [77]:
rice.loc[1:5,['receipt_id','receipts_item_id']]

receipt_id  receipts_item_id
1     9622257          32369294
2     9446359          31885876
3     9470290          31930241
4     9643416          32418582
5     9692093          32561236

In [83]:
rice.head()

receipt_id  receipts_item_id       purchase_time category sub_category  \
1     9622257          32369294 2018-07-22 21:19:00     Rice         Rice   
2     9446359          31885876 2018-07-15 16:17:00     Rice         Rice   
3     9470290          31930241 2018-07-15 12:12:00     Rice         Rice   
4     9643416          32418582 2018-07-24 08:27:00     Rice         Rice   
5     9692093          32561236 2018-07-26 11:28:00     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket      128000         0         1   2018-07  
2   minimarket      102750         0         1   2018-07  
3  supermarket       64000         0         3   2018-07  
4   minimarket       65000         0         1   2018-07  
5  supermarket      124500         0         1   2018-07

In [82]:
rice.iloc[0:5,0:4]

receipt_id  receipts_item_id       purchase_time category
1     9622257          32369294 2018-07-22 21:19:00     Rice
2     9446359          31885876 2018-07-15 16:17:00     Rice
3     9470290          31930241 2018-07-15 12:12:00     Rice
4     9643416          32418582 2018-07-24 08:27:00     Rice
5     9692093          32561236 2018-07-26 11:28:00     Rice

In [84]:
rice.loc[[1,3,5],:]

receipt_id  receipts_item_id       purchase_time category sub_category  \
1     9622257          32369294 2018-07-22 21:19:00     Rice         Rice   
3     9470290          31930241 2018-07-15 12:12:00     Rice         Rice   
5     9692093          32561236 2018-07-26 11:28:00     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket      128000         0         1   2018-07  
3  supermarket       64000         0         3   2018-07  
5  supermarket      124500         0         1   2018-07

In [87]:
rice.iloc[[0,2,4],:]

receipt_id  receipts_item_id       purchase_time category sub_category  \
1     9622257          32369294 2018-07-22 21:19:00     Rice         Rice   
3     9470290          31930241 2018-07-15 12:12:00     Rice         Rice   
5     9692093          32561236 2018-07-26 11:28:00     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
1  supermarket      128000         0         1   2018-07  
3  supermarket       64000         0         3   2018-07  
5  supermarket      124500         0         1   2018-07

### Bonus
Putting together what you've learned so far, use .iloc to subset the last 2 rows of our dataframe, and the first 4 columns. If you perform this exercise correctly, the output x should be a pandas dataframe (type(x) is a pandas DataFrame object)

Bonus: If you want an extra challenge, try and perform this operation three times;

- Use only .iloc[ , ]
- Use tail(2) to get the last 2 rows then chain it with .iloc
- Use rice.shape[0]-2: to get the last 2 rows in your .iloc operation

In [89]:
# code here
#cara 1
rice.iloc[len(rice)-2:len(rice),0:4]

receipt_id  receipts_item_id       purchase_time category
11999     5678748          17317935 2017-12-08 22:04:00     Rice
12000     5702411          17341559 2017-12-10 17:04:00     Rice

In [90]:
rice.iloc[11998:12000, 0:4]

receipt_id  receipts_item_id       purchase_time category
11999     5678748          17317935 2017-12-08 22:04:00     Rice
12000     5702411          17341559 2017-12-10 17:04:00     Rice

In [93]:
# line 2
rice.tail(2).iloc[:2, 0:4]

receipt_id  receipts_item_id       purchase_time category
11999     5678748          17317935 2017-12-08 22:04:00     Rice
12000     5702411          17341559 2017-12-10 17:04:00     Rice

In [94]:
#line2
rice.tail(2).iloc[:, 0:4]


receipt_id  receipts_item_id       purchase_time category
11999     5678748          17317935 2017-12-08 22:04:00     Rice
12000     5702411          17341559 2017-12-10 17:04:00     Rice

In [96]:
rice.iloc[(rice.shape[0]-2):, 0:4]

receipt_id  receipts_item_id       purchase_time category
11999     5678748          17317935 2017-12-08 22:04:00     Rice
12000     5702411          17341559 2017-12-10 17:04:00     Rice

In [ ]:
a 2 baris pertama
b 2 baris terakhir

In [ ]:
circular indexing 

In [20]:
rice.iloc[:-2,:]

receipt_id  receipts_item_id     purchase_time category sub_category  \
1         9622257          32369294   7/22/2018 21:19     Rice         Rice   
2         9446359          31885876   7/15/2018 16:17     Rice         Rice   
3         9470290          31930241   7/15/2018 12:12     Rice         Rice   
4         9643416          32418582    7/24/2018 8:27     Rice         Rice   
5         9692093          32561236   7/26/2018 11:28     Rice         Rice   
6         9504155          32030785   7/17/2018 18:05     Rice         Rice   
7         9822589          32935097   7/29/2018 18:18     Rice         Rice   
8         9706953          32593606   7/25/2018 12:48     Rice         Rice   
9         9699516          32573843   7/26/2018 16:41     Rice         Rice   
10        9444092          31913062   7/14/2018 21:17     Rice         Rice   
11        9198777          31125365    7/2/2018 15:39     Rice         Rice   
12        9763483          32856560    7/31/2018 5:51     Rice         Rice   
13        9691090          32552145   7/26/2018 11:43     Rice         Rice   
14        9624752          32369065   7/23/2018 14:22     Rice         Rice   
15        9543709          32142234   7/19/2018 18:05     Rice         Rice   
16        9763209          32744997   7/20/2018 12:53     Rice         Rice   
17        9481937          31964491   7/15/2018 19:10     Rice         Rice   
18        9818583          33018018   7/31/2018 15:34     Rice         Rice   
19        9362870          31602268   7/11/2018 17:16     Rice         Rice   
20        9343021          31535600    7/10/2018 7:29     Rice         Rice   
21        9627536          32372067   7/22/2018 20:04     Rice         Rice   
22        9372162          31632318    7/12/2018 6:37     Rice         Rice   
23        9289273          31380379    7/7/2018 15:17     Rice         Rice   
24        9767244          32763935    7/29/2018 8:25     Rice         Rice   
25        9604254          33196585    7/20/2018 7:41     Rice         Rice   
26        9262054          31558889    7/6/2018 21:47     Rice         Rice   
27        9645455          32424371   7/24/2018 10:45     Rice         Rice   
28        9215930          31170911    7/4/2018 19:18     Rice         Rice   
29        9597867          32294250   7/22/2018 10:51     Rice         Rice   
30        9197319          31118551    7/3/2018 21:42     Rice         Rice   
...           ...               ...               ...      ...          ...   
11969     5593860          17000940   12/1/2017 19:59     Rice         Rice   
11970     5600175          17002217   12/2/2017 17:03     Rice         Rice   
11971     5764051          17543485   12/16/2017 8:42     Rice         Rice   
11972     5962562          18384811  12/30/2017 19:14     Rice         Rice   
11973     5873850          17861252  12/24/2017 15:31     Rice         Rice   
11974     5826724          17716931  12/21/2017 12:06     Rice         Rice   
11975     5864736          18026296  12/23/2017 15:09     Rice         Rice   
11976     5778316          17562388  12/16/2017 17:30     Rice         Rice   
11977     5961252          18123996  12/31/2017 17:11     Rice         Rice   
11978     5698801          17326374  12/10/2017 16:14     Rice         Rice   
11979     5898495          17935663  12/26/2017 18:49     Rice         Rice   
11980     5976949          18168553  12/31/2017 11:33     Rice         Rice   
11981     5791408          17631871  12/17/2017 11:59     Rice         Rice   
11982     5856357          17805171  12/23/2017 18:10     Rice         Rice   
11983     5638097          17128628    12/5/2017 8:00     Rice         Rice   
11984     5789102          17619822  12/17/2017 15:35     Rice         Rice   
11985     5671322          17251772   12/8/2017 12:01     Rice         Rice   
11986     5894948          17931856  12/26/2017 16:22     Rice         Rice   
11987     5920761          18017126  12/28/2017 17:22    

### Dive Deeper:

In the following code, we read in `companies.csv`. Take a peek at the data using `head` or `tail`. 

Perform a *label*-based or *integer*-based indexing (whichever deemed more appropriate) by subsetting for the row corresponding to 'Li and Partners' and tries to answer the following questions:

1. Where is *Li and Partners* located?
2. How much `Returns` did PT. Algoritma Data Indonesia make?
3. What about the `Forecasted Growth` for Palembang Konsultansi?

In [99]:
# code here
companies = pd.read_csv('data_input/companies.csv')
companies

ID                 Customer Name Consulting Sales Software Sales  \
0  30940               New Media Group       IDR7125000     IDR5500000   
1  82391               Li and Partners        IDR420000      IDR820000   
2  18374      PT. Kreasi Metrik Solusi                0      IDR550403   
3  57531  PT. Algoritma Data Indonesia        IDR850000      IDR395500   
4  19002         Palembang Konsultansi       IDR2115000              0   
5  31142     PT. Surya Citra Manajemen        IDR960000      IDR503000   

  Forecasted Growth       Returns  Month  Day  Year  Location     Account  
0            30.00%  IDR1,500,000      1   10  2017   Jakarta  Enterprise  
1            10.00%    IDR400,000      6   15  2016   Jakarta     Startup  
2            25.00%             0      3   29  2012  Surabaya  Enterprise  
3             4.00%             0      7   17  2017   Jakarta     Startup  
4           -15.00%             0      2   24  2018   Bandung     Startup  
5            19.00%             0      1   19  2019   Jakarta  Enterprise

In [102]:
companies.loc[1,['Location']]

Location    Jakarta
Name: 1, dtype: object

In [104]:
companies.loc[3,['Returns']]

Returns    0
Name: 3, dtype: object

In [105]:
companies.iloc[4,4]

'-15.00%'

In [107]:
companies.iloc[3:4,9:10]

Location
3  Jakarta

In [108]:
companies.loc[[3],['Returns']]

Returns
3       0

## Conditional Subsetting

Selain menggunakan operator dan method yang telah dibahas sebelumnya seperti `loc` dan `iloc`, kita dapat melakukan subsetting berdasarkan kondisi tertentu. Misal pada dataframe `rice` yang telah kita buat sebelumnya, kita ingin mengambil beberapa data dengan kondisi sebagai berikut:
- Kita akan mengambil data yang transaksinya berasal dari supermarket -> `.format=='supermarket'`
- Kita akan mengambil transaksi produk yang harganya lebih dari 10000 -> `.unit_price>10000`
- Kita hanya akan mengambil semua transaksi yang pembelian barangnya tidak sama dengan 0 -> `.quantity!=0`, dll

In [16]:
rice = pd.read_csv("data_input/rice.csv", index_col=1)
rice = rice.drop('Unnamed: 0', axis=1)
rice.head(3)

receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9446359             31885876  7/15/2018 16:17     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000         0         1   2018-07  
9446359      minimarket  102750.000         0         1   2018-07  
9470290     supermarket   64000.000         0         3   2018-07

Syntax penulisan untuk conditional subsetting adalah:

**```
df[df['column_name']=='condition']
```**

In [28]:
rice[rice.['format pasar'] == 'supermarket'].head()

receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   
9692093             32561236  7/26/2018 11:28     Rice         Rice   
9822589             32935097  7/29/2018 18:18     Rice         Rice   
9444092             31913062  7/14/2018 21:17     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000         0         1   2018-07  
9470290     supermarket   64000.000         0         3   2018-07  
9692093     supermarket  124500.000         0         1   2018-07  
9822589     supermarket   66500.000         0         1   2018-07  
9444092     supermarket   64000.000         0         3   2018-07

In [34]:
rice.discount = rice.discount.astype('object')
rice.quantity = rice.quantity.astype('object')

In [33]:
rice.dtypes

receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount             object
quantity             object
yearmonth            object
sum                  object
dtype: object

In [35]:
alagoritma = "programming"
algoritma2 = "data analytics"
alagoritma+algoritma2

'programmingdata analytics'

In [ ]:
sum = 10
# int-> 1
# object 

In [32]:
rice['sum'] =  rice['quantity'] + rice['discount']
rice.head()

receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9446359             31885876  7/15/2018 16:17     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   
9643416             32418582   7/24/2018 8:27     Rice         Rice   
9692093             32561236  7/26/2018 11:28     Rice         Rice   

                 format  unit_price discount quantity yearmonth sum  
receipt_id                                                           
9622257     supermarket  128000.000        0        1   2018-07   1  
9446359      minimarket  102750.000        0        1   2018-07   1  
9470290     supermarket   64000.000        0        3   2018-07   3  
9643416      minimarket   65000.000        0        1   2018-07   1  
9692093     supermarket  124500.000        0        1   2018-07   1

In [23]:
## code here
rice[rice['format']=='supermarket'].head()


receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   
9692093             32561236  7/26/2018 11:28     Rice         Rice   
9822589             32935097  7/29/2018 18:18     Rice         Rice   
9444092             31913062  7/14/2018 21:17     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000         0         1   2018-07  
9470290     supermarket   64000.000         0         3   2018-07  
9692093     supermarket  124500.000         0         1   2018-07  
9822589     supermarket   66500.000         0         1   2018-07  
9444092     supermarket   64000.000         0         3   2018-07

Kita juga dapat menggunakan `&` dan `|` untuk melakukan subsetting lebih dari 1 kondisi. Misalnya kita ingin melihat data penjualan dari seorang pegawai yang jumlahnya lebih dari 5000, maka kita dapat menggunakan syntax:
```
sales[(sales.salesperson == 'Moana') | (sales.amount > 5000)]
```
Untuk subsetting dengan kondisi lebih dari 1, setiap kondisi diletakkan di dalam tanda kurung `()` atau bisa ditulis dengan syntax berikut:
```
df[(df['column_name']=='condition 1') operator (df['column_name']=='condition 2')]
```

In [26]:
#code here
#untuk 2 kondisi 

rice[(rice['format']=='supermarket')&(rice['unit_price']>20000)].shape[1]

9

### Knowledge Check
In the cell below, write code using conditional subsetting and answer the following questions:

Say as a company, we define "bulk purchase transaction" to be any transaction containing at least 8 units of the same item (quantity column). In our dataset, 
- How many "bulk purchases" do we have?
- How many transactions in our dataset took place in a minimarket?
- How many transactions in our dataset took place in a hypermarket and has at least 10 units (quantity column)?
Tip: You may find the .shape attribute convenient in extracting the number of rows / columns from a dataframe

In [40]:
bulk = rice[(rice.quantity >= 8)]
bulk.shape[0]

61

In [36]:
## code here
bulk = rice[rice['quantity']>=8]
bulk.shape[0]


61

In [37]:
rice[rice['format']=='minimarket'].shape[0]

7088

In [38]:
rice[(rice['format']=='hypermarket')&(rice['quantity']>=10)].shape[0]

8

**Bonus Challenge**

Dari penjelasan di atas, bagaimana cara Anda  melakukan subsetting pada dataframe rice untuk semua transaksi yang dilakukan baik di minimarket atau supermarket pada bulan Juli di tahun 2018? Ada berapa transaksi yang terjadi?

In [17]:
#code here
sales_july = rice[((rice['format']=='supermarket')|(rice['format']=='minimarket'))&\
                  (rice['yearmonth']=='2018-07')]
sales_july.shape

(935, 9)

In [18]:
sales_july.format.unique()

array(['supermarket', 'minimarket'], dtype=object)

In [19]:
sales_july.yearmonth.unique()

array(['2018-07'], dtype=object)

Kesimpulan : Ada 935 transaksi yang dilakukan di supermarket atau minimarket pada bulan July 2018.

### Additional read_csv()
- comment
- nrows
- skiprows

Case : analytics.csv

In [43]:
analytic = pd.read_csv('data_input/analytics.csv', comment='#').head(11)
analytic

Language   Users New Users Sessions Bounce Rate  Pages / Session  \
0     en-us  23,497    22,696   35,113       0.54%            6.350   
1     id-id   7,797     7,613   10,617       0.65%            4.890   
2     en-gb   3,198     3,086    4,625       0.56%            5.520   
3        id   2,171     2,094    2,724       0.29%            4.750   
4        en     224       214      295       3.05%            5.030   
5     id-us     204       202      235       1.28%            3.640   
6     th-th     186       186      224       0.45%            3.500   
7     en-sg     132       125      231       0.00%            6.920   
8     en-id     126       119      171       0.00%            6.390   
9     en-au      93        91      127       0.00%            5.560   
10      NaN  38,351    37,140   55,195       0.63%            5.860   

   Avg. Session Duration Goal Conversion Rate Goal Completions Goal Value  
0               00:02:32               27.21%            9,555      $0.00  
1               00:01:38               16.26%            1,726      $0.00  
2               00:01:50               21.36%              988      $0.00  
3               00:01:43               17.18%              468      $0.00  
4               00:01:44               26.78%               79      $0.00  
5               00:00:44                4.68%               11      $0.00  
6               00:00:40                4.91%               11      $0.00  
7               00:02:10               39.83%               92      $0.00  
8               00:04:10               26.32%               45      $0.00  
9               00:01:22               19.69%               25      $0.00  
10              00:02:13               23.96%           13,222      $0.00

In [44]:
analytic = pd.read_csv('data_input/analytics.csv', comment='#', nrows=11)
analytic

Language   Users New Users Sessions Bounce Rate  Pages / Session  \
0     en-us  23,497    22,696   35,113       0.54%            6.350   
1     id-id   7,797     7,613   10,617       0.65%            4.890   
2     en-gb   3,198     3,086    4,625       0.56%            5.520   
3        id   2,171     2,094    2,724       0.29%            4.750   
4        en     224       214      295       3.05%            5.030   
5     id-us     204       202      235       1.28%            3.640   
6     th-th     186       186      224       0.45%            3.500   
7     en-sg     132       125      231       0.00%            6.920   
8     en-id     126       119      171       0.00%            6.390   
9     en-au      93        91      127       0.00%            5.560   
10      NaN  38,351    37,140   55,195       0.63%            5.860   

   Avg. Session Duration Goal Conversion Rate Goal Completions Goal Value  
0               00:02:32               27.21%            9,555      $0.00  
1               00:01:38               16.26%            1,726      $0.00  
2               00:01:50               21.36%              988      $0.00  
3               00:01:43               17.18%              468      $0.00  
4               00:01:44               26.78%               79      $0.00  
5               00:00:44                4.68%               11      $0.00  
6               00:00:40                4.91%               11      $0.00  
7               00:02:10               39.83%               92      $0.00  
8               00:04:10               26.32%               45      $0.00  
9               00:01:22               19.69%               25      $0.00  
10              00:02:13               23.96%           13,222      $0.00

In [47]:
analytic = pd.read_csv('data_input/analytics.csv', skiprows=18)
analytic.head()

Day Index Users
0    1/1/19   122
1    1/2/19   174
2    1/3/19   240
3    1/4/19   231
4    1/5/19   172

## Referencing and Copying
Pada python, tanda sama dengan (=) bisa digunakan untuk merefer pada sebuah Dataframe. Sekarang mari kita lihat kembali data `rice.csv`:

In [10]:
rice = pd.read_csv("data_input/rice.csv", index_col=1)
rice = rice.drop('Unnamed: 0', axis=1)
rice.head(3)

receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9446359             31885876  7/15/2018 16:17     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000         0         1   2018-07  
9446359      minimarket  102750.000         0         1   2018-07  
9470290     supermarket   64000.000         0         3   2018-07

Selanjutnya, kita buat dataframe baru dengan nama `rice_july` yang menyimpan transaksi di bulan July 2018. Kemudian selama periode tersebut kita memberikan diskon 15 persen di semua produk, sehingga kita  juga harus mengganti semua nilai di `rice_july.discount` menjadi 15.

In [42]:
#code here
rice_july = rice
rice_july['discount'] = 15
rice.head()


receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9446359             31885876  7/15/2018 16:17     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   
9643416             32418582   7/24/2018 8:27     Rice         Rice   
9692093             32561236  7/26/2018 11:28     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000        15         1   2018-07  
9446359      minimarket  102750.000        15         1   2018-07  
9470290     supermarket   64000.000        15         3   2018-07  
9643416      minimarket   65000.000        15         1   2018-07  
9692093     supermarket  124500.000        15         1   2018-07

In [39]:
rice = pd.read_csv("data_input/rice.csv", index_col=1)
rice = rice.drop('Unnamed: 0', axis=1)
rice.head(3)

receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9446359             31885876  7/15/2018 16:17     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000         0         1   2018-07  
9446359      minimarket  102750.000         0         1   2018-07  
9470290     supermarket   64000.000         0         3   2018-07

In [44]:
rice_july = rice.copy()
rice_july['discount'] = 15
rice.head()

receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9446359             31885876  7/15/2018 16:17     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   
9643416             32418582   7/24/2018 8:27     Rice         Rice   
9692093             32561236  7/26/2018 11:28     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000         0         1   2018-07  
9446359      minimarket  102750.000         0         1   2018-07  
9470290     supermarket   64000.000         0         3   2018-07  
9643416      minimarket   65000.000         0         1   2018-07  
9692093     supermarket  124500.000         0         1   2018-07

In [45]:
rice_july.head()

receipts_item_id    purchase_time category sub_category  \
receipt_id                                                            
9622257             32369294  7/22/2018 21:19     Rice         Rice   
9446359             31885876  7/15/2018 16:17     Rice         Rice   
9470290             31930241  7/15/2018 12:12     Rice         Rice   
9643416             32418582   7/24/2018 8:27     Rice         Rice   
9692093             32561236  7/26/2018 11:28     Rice         Rice   

                 format  unit_price  discount  quantity yearmonth  
receipt_id                                                         
9622257     supermarket  128000.000        15         1   2018-07  
9446359      minimarket  102750.000        15         1   2018-07  
9470290     supermarket   64000.000        15         3   2018-07  
9643416      minimarket   65000.000        15         1   2018-07  
9692093     supermarket  124500.000        15         1   2018-07

In [46]:
rice_july.to_csv('data_input/rice_july.csv')